packages/dht/src/connection/ConnectionManager.ts

```diff
@@ -0,0 +1,676 @@
+            throw new Err.CouldNotStart(`Cannot start already ${this.state} module`)
+        }
+        this.state = ConnectionManagerState.RUNNING
+        logger.trace(`Starting ConnectionManager...`)
+        // Garbage collection of connections
+        this.disconnectorIntervalRef = setInterval(() => {
+            logger.trace('disconnectorInterval')
+            const LAST_USED_LIMIT = 20000
+            this.garbageCollectConnections(this.config.maxConnections, LAST_USED_LIMIT)
+        }, 5000)
+        if (!this.config.simulator) {
+            await this.webSocketConnector!.start()
+            const connectivityResponse = await this.webSocketConnector!.checkConnectivity()
+            const ownPeerDescriptor = peerDescriptorGeneratorCallback!(connectivityResponse)
+            this.ownPeerDescriptor = ownPeerDescriptor
+            this.webSocketConnector!.setOwnPeerDescriptor(ownPeerDescriptor)
+            this.webrtcConnector!.setOwnPeerDescriptor(ownPeerDescriptor)
+        }
+    }
+
+    public async stop(): Promise<void> {
+        if (this.state === ConnectionManagerState.STOPPED || this.state === ConnectionManagerState.STOPPING) {
+            return
+        }
+        this.state = ConnectionManagerState.STOPPING
+        logger.trace(`Stopping ConnectionManager`)
+        if (this.disconnectorIntervalRef) {
+            clearInterval(this.disconnectorIntervalRef)
+        }
+        if (!this.config.simulator) {
+            await this.webSocketConnector!.stop()
+            this.webSocketConnector = undefined
+            await this.webrtcConnector!.stop()
+            this.webrtcConnector = undefined
+        } else {
+            await this.simulatorConnector!.stop()
+            this.simulatorConnector = undefined
+        }
+
+        await Promise.all(Array.from(this.connections.values()).map((peer) => {
+            return new Promise<void>((resolve, _reject) => {
+
+                if (peer.isHandshakeCompleted()) {
+
+                    this.gracefullyDisconnectAsync(peer.getPeerDescriptor()!, DisconnectMode.LEAVING)
+                        .then(() => { resolve() })
+                        .catch((e) => {
+                            logger.error(e)
+                            resolve()
+                        })
+                } else {
+                    logger.trace('handshake of connection not completed, force-closing')
+
+                    waitForEvent3<ManagedConnectionEvents>(peer!, 'disconnected', 2000)
+                        .then(() => {
+                            logger.trace('resolving after receiving disconnected event from non-handshaked connection')
+                            resolve()
+                        })
+                        .catch((e) => {
+                            logger.trace('force-closing non-handshaked connection timed out ' + e)
+                            resolve()
```

Comment:
This promise handling could be simpler with `await`? There is also some path handled only partially, as we get eslint warning "Each then() should return a value or throw"